### PR TITLE
Fix: isnan/isfinite

### PR DESCRIPTION
**Pull request by tkukurin**
I think this fixes a probable bug where invalid (`inf`) values could enter the generation procedure.
Example to generate about 1024 `inf` datasets:

```python
import torch, numpy as np
from tabicl.prior import PriorDataset
from tabicl.prior._mlp_scm import MLPSCM
from tabicl.prior._tree_scm import TreeSCM

N = 100_000  # num draws

def _f(self):  # same code as forward in the prior package
    causes = self.xsampler.sample()  # (seq_len, num_causes)
    outputs = [causes]
    for layer in self.layers:
        outputs.append(layer(outputs[-1]))
    outputs = outputs[2:]  # Start from 2 because the first layer is only linear without activation
    X, y = self.handle_outputs(causes, outputs)

    # HERE we introduced the change
    if torch.any(torch.isnan(X)) or torch.any(torch.isnan(y)):
        X[:] = 0.0
        y[:] = -100.0

    if self.num_outputs == 1:
        y = y.squeeze(-1)
    return X, y


# monkeypatched for easy check
MLPSCM.forward = _f
TreeSCM.forward = _f

np.random.seed(0); torch.manual_seed(0)
prior = PriorDataset(
    regression=True, batch_size=8, batch_size_per_gp=8,
    batch_size_per_subgp=8, min_features=2, max_features=64,
    max_seq_len=1024, min_train_size=0.25, max_train_size=0.9,
    prior_type="mlp_scm", device="cpu", n_jobs=1)

for i, batch in enumerate(prior):
    print(f"\r{i}", end="", flush=True)
    if i >= N: break
    X, y, d, _, _ = batch
    if not torch.isfinite(y).all():
        n_inf = (~torch.isfinite(y)).sum().item()
        bad_tables = (~torch.isfinite(y)).any(1)
        print(f"{i=}: {n_inf=}")
        for t in range(len(y)):
            if not torch.isfinite(y[t]).all():
                n_nan = torch.isnan(y[t]).sum().item()
                n_inf = torch.isinf(y[t]).sum().item()
                print(f"{t=}: {n_nan=}, {n_inf=}, {y.shape[1]=}")
        break
```

### CODE REVIEW

**Comment by ogrisel** [comment]
Thanks. I am not yet familiar with the prior code base so I will let @dholzmueller do the review of this PR.

Note that it would help to have a few automated tests to check for the expected invariants of the prior code ;)

**Comment by tkukurin** [comment]
sure, can you advise what/where you'd do it? 
if  @dholzmueller  approves I can add a test or just convert my PR description block into a file.

@dholzmueller 
the paper doesn't explicitly mention how to handle inf's 
but semantically I don't see how it'd make sense in the training procedure


P.S.

> familiar with the prior code base

no that's another company 😄


(bad joke sorry)

**Comment by dholzmueller** [comment]
Thanks! Did those infs actually occur and what happens in this case? I'm a bit hesitant to change the original prior, since people may want to use it to reproduce the TabICLv1 model.

**Comment by tkukurin** [comment]
*setup*:
generate synthetic data and train the model
*problem*:
can't have valid gradients with infty inputs, mainly wasted compute

entriely possible I was doing something wrong 
(based on your reaction no? it's basically all in that code block from my PR)

this is how prior does it: https://github.com/PriorLabs/TabPFN/blob/8f2d3ce50f208768f05981ce4ebe10606ce9cb7e/src/tabpfn/architectures/tabpfn_v3.py#L2306-L2312

afaik an equivalent doesn't exist for tabicl (i.e. the v2 paper iirc explicitly mentions not special casing nan/inf) 
unless some other training wrapper I missed handles this

redid a quick search again, there is some adhoc fix in softmax but no idea whose impl that would be (or Claude artistic freedom):
https://github.com/search?q=repo%3Asoda-inria%2Ftabicl+isfinite&type=code

equivalent might be to
* implement a filter elsewhere and leave the prior as-is 
* disregard entirely, maybe add comment about this behavior inline
 
but I'll leave that up to you, I'm fine closing this PR let me know

EDIT: I further assume if they did intend to perform infty filtering that might be described in appendix E

I don't remember whether it's addressed off the top of my head. if you are in parallel training other models (e.g. ones that expect to behave nicely w/ infty inputs) then it makes sense to abandon this PR probably

**Comment by dholzmueller** [comment]
> can't have valid gradients with infty inputs, mainly wasted compute

So what happened in this case, did the gradients get discarded by the gradient clipping or did this mess up the whole pre-training? And do you have an idea how often this occurs?

**Comment by tkukurin** [comment]
per PR description, frequency seems 1024/100k in the exact snippet I shared.
I assume freq is consistent but depends on feats/rows generated.

clipping impacts only grad I think, what I was seeing is likley `nan`s due to $x \cdot \frac{\partial f_w(x)}{\partial w}$.

didn't check in detail though, applied this fix instead due to other time constraints so this is all I got sorry.

**Comment by dholzmueller** [comment]
I would merge if
- there is a flag for disabling this behavior (to reproduce the original)
- the flag isn't set by default but it is noted in the TabICLv1 pre-training scripts that it can be used to reproduce the old behavior.

Could you do that or should I push something to your branch?

**Comment by tkukurin** [comment]
feel free to push whenever you find the time, don't think I'll be able to update soon
